ui/StatusQ/include/StatusQ/writableproxymodel.h

```diff
@@ -0,0 +1,96 @@
+    //Returns a VariantMap of the data at the given index
+    //The map contains the role names as keys and the data as values
+    Q_INVOKABLE QVariantMap get(int at) const;
+    //Sets the data at the given index
+    //The map contains the role names as keys and the data as values
+    Q_INVOKABLE bool set(int at, const QVariantMap& data);
+
+    bool dirty() const;
+
+    //QAbstractProxyModel overrides
+    void setSourceModel(QAbstractItemModel* sourceModel) override;
+
+    int	columnCount(const QModelIndex& parent = QModelIndex()) const override;
+    int	rowCount(const QModelIndex& parent = QModelIndex()) const override;
+
+    QVariant data(const QModelIndex& index, int role = Qt::DisplayRole) const override;
+    bool setData(const QModelIndex& index, const QVariant& value, int role = Qt::EditRole) override;
+
+    QMap<int, QVariant> itemData(const QModelIndex& index) const override;
+    bool setItemData(const QModelIndex& index, const QMap<int, QVariant>& roles) override;
+
+    bool removeRows(int row, int count, const QModelIndex& parent = QModelIndex()) override;
+    bool insertRows(int row, int count, const QModelIndex& parent = QModelIndex()) override;
+
+    QModelIndex index(int row, int column, const QModelIndex& parent = QModelIndex()) const override;
+    QModelIndex sibling(int row, int column, const QModelIndex& idx) const override;
+    QModelIndex parent(const QModelIndex& child) const override;
+    QModelIndex mapToSource(const QModelIndex& proxyIndex) const override;
+    QModelIndex mapFromSource(const QModelIndex& sourceIndex) const override;
+
+    bool hasChildren(const QModelIndex& parent = QModelIndex()) const override;
+    void revert() override;
+
+    // TODO: implement these
+    // bool moveRows(const QModelIndex& sourceParent, int sourceRow, int count, const QModelIndex& destinationParent, int destinationChild) override;
+    // bool	submit() override;
+
+signals:
+    void dirtyChanged();
+
+private:
+    void setDirty(bool flag);
+
+    void onSourceDataChanged(const QModelIndex& topLeft, const QModelIndex& bottomRight, const QVector<int>& roles);
+    void onRowsAboutToBeInserted(const QModelIndex& parent, int start, int end);
+    void onRowsInserted(const QModelIndex& parent, int first, int last);
+    void onRowsAboutToBeRemoved(const QModelIndex& parent, int start, int end);
+    void onRowsRemoved(const QModelIndex& parent, int first, int last);
+    void onModelAboutToBeReset();
+    void onModelReset();
+    void onLayoutAboutToBeChanged(const QList<QPersistentModelIndex>& sourceParents, QAbstractItemModel::LayoutChangeHint hint);
+    void onLayoutChanged(const QList<QPersistentModelIndex>& sourceParents, QAbstractItemModel::LayoutChangeHint hint);
+    void onRowsMoved(const QModelIndex& sourceParent, int sourceStart, int sourceEnd, const QModelIndex& destinationParent, int destinationRow);
+
+    bool m_dirty{false};
+
+    QScopedPointer<WritableProxyModelPrivate> d;
```

Comment:
We don't need it. I've preferred it just to have a clean header as it's going to be the one everyone reads in most cases. Removing it is a quick job and I can do it if you think it would help.

Reply:
In the mean time I've moved the `m_dirty` to private part.

Reply:
Imo public/private sections are separated enough and all cpp types we have are done this way. But if you like this approach feel free to keep it as it is :)